src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,145 @@
+
+        public Node(T element) {
+            this.element = element;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+        if (size == 0) {
+            head = tail = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        Node<T> newNode = new Node<>(value);
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index " + index + " is out of list bound");
+        }
+
+        if (head == null) {
+            head = tail = newNode;
+        } else if (index == 0) {
+            newNode.next = head;
+            head.prev = newNode;
+            head = newNode;
+        } else if (index == size) {
+            add(value);
+            return;
+        } else {
+            Node<T> currentNode = getNodeByIndex(index);
+            Node<T> prevNode = currentNode.prev;
+            newNode.next = currentNode;
+            currentNode.prev = newNode;
+            newNode.prev = prevNode;
+            prevNode.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexRange(index);
+        return getNodeByIndex(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexRange(index);
+        Node<T> currentNode = getNodeByIndex(index);
+        T changedElement = currentNode.element;
+        currentNode.element = value;
+        return changedElement;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexRange(index);
+        Node<T> removedNode = getNodeByIndex(index);
+        T value = null;
+
+        if (index == 0) {
+            value = removedNode.element;
+            head = removedNode.next;
+            if (head == null) {
+                tail = null;
+            }
+        } else if (index == size - 1) {
+            value = removedNode.element;
+            tail = removedNode.prev;
+            removedNode.prev = null;
+            tail.next = null;
+        } else {
+            value = removedNode.element;
+            Node<T> prevNode = removedNode.prev;
+            Node<T> nextNode = removedNode.next;
+            prevNode.next = nextNode;
+            nextNode.prev = prevNode;
+        }
+        size--;
+        return value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < size; i++) {
+            if (currentNode.element != null && currentNode.element.equals(object)
+                    || currentNode.element == object) {
+                remove(i);
```

Comment:
```suggestion
                unlink(currentNode);
```
<img width="1077" alt="image" src="https://user-images.githubusercontent.com/81781479/202228676-6ea403ac-b5ab-45ab-880f-379d1fa746cb.png">